### PR TITLE
Support IAM auth in remaining AWS sources

### LINKED ISSUE #791: Add support for iamRole in all AWS sources
following up on https://github.com/triggermesh/triggermesh/pull/783

we need to add this iamRole authentication support in all AWS sources.

### PR DESCRIPTION
**Pull request by antoineco**
Identical to #783
Closes #791

Note: omit source for Amazon SNS, which is unfit for this authentication method due to its multi-tenant nature.

### CODE REVIEW

File: pkg/apis/sources/v1alpha1/awss3_types.go [hunk 1]
```diff
@@ -67,11 +67,10 @@ type AWSS3SourceSpec struct {
 	// https://docs.aws.amazon.com/AmazonS3/latest/userguide/notification-how-to-event-types-and-destinations.html
 	EventTypes []string `json:"eventTypes"`
 
-	// The destination of notifications originating from the Amazon S3 bucket.
-	//
+	// The intermediate destination of notifications originating from the
+	// Amazon S3 bucket, before they are retrieved by TriggerMesh.
```
Comment (odacremolbap) [inline pkg/apis/sources/v1alpha1/awss3_types.go:71]: same as above
```suggestion
	// Amazon S3 bucket, before they are retrieved by this source.
```

**Review by odacremolbap — APPROVED** [review]
Made a comment, although I am unsure if it is a valid one.

File: config/300-awss3source.yaml [hunk 2]
```diff
@@ -151,12 +151,14 @@ spec:
                   - s3:Replication:OperationMissedThreshold
                   - s3:Replication:OperationReplicatedAfterThreshold
               destination:
-                description: The destination of notifications originating from the Amazon S3 bucket. If omitted, an
-                  Amazon SQS queue is automatically created and associated with the bucket.
+                description: The intermediate destination of notifications originating from the Amazon S3 bucket, before
+                  they are retrieved by TriggerMesh. If omitted, an Amazon SQS queue is automatically created and
```
[the remark this answers is not in the document]
Comment (antoineco) [reply config/300-awss3source.yaml:155]: Sounds valid to me 👍 

This wording is used verbatim in almost every GCP and Azure source too. Do you mind if I submit that change separately? (Right after this PR, promise)

An example:

https://github.com/triggermesh/triggermesh/blob/f50188dd20be6f37810a8803ecccf8bd82feca3b/config/300-azureactivitylogssource.yaml#L118

[the remark this answers is not in the document]
Comment (odacremolbap) [reply config/300-awss3source.yaml:155]: of course!

[the remark this answers is not in the document]
Comment (antoineco) [reply config/300-awss3source.yaml:155]: Submitted in #836

**Comment by vashibhavin** [comment]
I am a little new the Knative ecosystem. I am trying to setup Kinesis data source and I just realized that there are 2 different event sources available. 
https://github.com/triggermesh/aws-event-sources/blob/main/config/300-awskinesis.yaml and https://github.com/triggermesh/triggermesh/blob/main/config/300-awskinesissource.yaml It took me a while to realize that these were 2 different projects. I am looking for support for IAM Auth specifically and it seems to be supported by the Event source in the triggermesh repository. Which event source should I be using?

**Comment by cab105** [comment]
Hi @vashibhavin. The aws-event-sources repo was folded into the triggermesh/triggermesh repo quite a while ago so this is the correct place to look.

**Comment by sebgoa** [comment]
The original work was done in https://github.com/triggermesh/aws-event-sources but this repo is archived.

You should indeed use the latest source available here.

There is a sample configuration for the Kinesis source: https://github.com/triggermesh/triggermesh/blob/main/config/samples/sources/awskinesissource.yaml

Indeed the authentication mechanism can be a `iamRole`

**Comment by sudughonge** [comment]
Hi @antoineco I've gone through some of the issues and PRs on this repo and you've explained in most of them at adding `iamRole` authentication for SNS is not possible. Could you please elaborate why? As someone relatively new to this, SNS and SQS look similar to me, so it would greatly help in getting a more detailed explanation about their differences, atleast wrt triggermesh.

**Comment by antoineco** [comment]
@sudughonge it's because the SNS source spawns a web server that acts as a webhook for receiving notifications from SNS, and that webhook is _shared between all instances of the SNS source_.
(We called that a **multi-tenant** source adapter.)

For example, if you deploy `source1` and `source2` (inside the same namespace), they will _both_ be backed by the same `awssnssource-adapter` webhook.

This doesn't work with the IAM role authentication if both sources want to use a different IAM role, because that mode of authentication only allows providing _one_ IAM role per Kubernetes Pod.

With that being said, this design could be revisited to not make this source "multi-tenant".

**Comment by sudughonge** [comment]
@antoineco I see, thanks for explaining. I'm assuming then, by design, this is not the case for SQS, i.e., each SQS source spawns its own webhook, and therefore doesn't run into the "multi-tenant" problem??

**Comment by antoineco** [comment]
Correct, the SQS source is single-tenant (one source adapter per source instance).

**Comment by sudughonge** [comment]
Is making SNS single-tenant a complicated task? I'd be happy to contribute if this needs person power, and if you can point me to where I can start looking.

**Comment by antoineco** [comment]
Not excessively complicated I believe.

- **Controller code**
    Simply remove this: https://github.com/triggermesh/triggermesh/blob/77169bfdafc2c54b6c5a6db91a1a09bf479d2a7e/pkg/apis/sources/v1alpha1/awssns_lifecycle.go#L73-L76
- **Adapter code**
    Use Main instead of MainWithController (take example on the SQS source): https://github.com/triggermesh/triggermesh/blob/77169bfdafc2c54b6c5a6db91a1a09bf479d2a7e/cmd/awssnssource-adapter/main.go#L24-L26
    Remove all the controller related code: https://github.com/triggermesh/triggermesh/blob/77169bfdafc2c54b6c5a6db91a1a09bf479d2a7e/pkg/sources/adapter/awssnssource/controller.go#L32-L55
    https://github.com/triggermesh/triggermesh/blob/77169bfdafc2c54b6c5a6db91a1a09bf479d2a7e/pkg/sources/adapter/awssnssource/reconciler.go#L35-L103
    From there, adapt `pkg/sources/adapter/awssnssource/adapter.go`. Again, you can use other sources as example.

**Comment by sudughonge** [comment]
Thanks! This helps. I'll take a stab at this today.

**Comment by sebgoa** [comment]
@sudughonge if you join us on Slack https://join.slack.com/t/triggermesh-community/shared_invite/zt-1kngevosm-MY7kqn9h6bT08hWh8PeltA  the team can guide you through making that PR.

**Pull request merged**